src/mirror.rs

```diff
@@ -1,21 +1,68 @@
-use crate::{fetch, util, Ctx, Krate, Source};
+use crate::{fetch, Ctx, Krate, Registry, Source};
+use anyhow::Context;
 use anyhow::Error;
-use std::{convert::TryFrom, time::Duration};
+use futures::StreamExt;
+use std::path::Path;
+use std::time::Duration;
 use tracing::{debug, error, info};
 use tracing_futures::Instrument;
 
-pub async fn registry_index(backend: crate::Storage, max_stale: Duration) -> Result<usize, Error> {
-    let url = url::Url::parse("git+https://github.com/rust-lang/crates.io-index.git")?;
-    let canonicalized = util::Canonicalized::try_from(&url)?;
-    let ident = canonicalized.ident();
+pub async fn registries_index(
+    backend: crate::Storage,
+    max_stale: Duration,
+    registries: Vec<Registry>,
+) -> Result<usize, Error> {
+    let bytes = futures::stream::iter(registries)
+        .map(|registry| {
+            let backend = backend.clone();
+            let index = registry.index.clone();
+            async move {
+                let res: Result<usize, Error> = registry_index(backend, max_stale, &registry)
+                    .instrument(tracing::debug_span!("upload registry"))
+                    .await;
+                res
+            }
+            .instrument(tracing::debug_span!("mirror registry", %index))
+        })
+        .buffer_unordered(32);
+    let total_bytes = bytes
+        .fold(0usize, |acc, res| async move {
+            match res {
+                Ok(a) => a + acc,
+                Err(e) => {
+                    error!("{:#}", e);
+                    acc
+                }
+            }
+        })
+        .await;
+    Ok(total_bytes)
+}
 
+pub async fn registry_index(
+    backend: crate::Storage,
+    max_stale: Duration,
+    registry: &Registry,
+) -> Result<usize, Error> {
+    let url = url::Url::parse(&registry.index)?;
+    let ident = registry.short_name()?;
+
+    let path = Path::new(url.path());
+    let name = if path.ends_with(".git") {
+        path.file_stem().context("failed to get registry name")?
+    } else {
+        path.file_name().context("failed to get registry name")?
+    };
     // Create a fake krate for the index, we don't have to worry about clashing
     // since we use a `.` which is not an allowed character in crate names
     let krate = Krate {
-        name: "crates.io-index".to_owned(),
+        name: String::from(
+            name.to_str()
+                .context("failed conversion from OsStr to String")?,
+        ),
```

Comment:
We can't rely on the path of the registry to not clash with an actual crate name, it should be safe to just use the short ident for the registry instead.